src/Std/Time/DateTime/Timestamp.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sofia Rodrigues
+-/
+prelude
+import Std.Time.Internal
+import Init.Data.Int
+import Std.Time.Time
+import Std.Time.Date
+
+namespace Std
+namespace Time
+open Internal
+
+set_option linter.all true
+
+/--
+Represents a point in time relative to the UNIX Epoch, with nanoseconds precision.
+-/
+structure Timestamp where
+
+  /--
+  Second offset of the timestamp.
+  -/
+  second : Second.Offset
+
+  /--
+  Nanosecond span that ranges from -999999999 and 999999999
+  -/
+  nano : Nanosecond.Span
+
+  /--
+  Proof that the timestamp is valid, ensuring that the `second` and `nano` values are correctly related.
+  -/
+  proof : (second.val ≥ 0 ∧ nano.val ≥ 0) ∨ (second.val ≤ 0 ∧ nano.val ≤ 0)
+  deriving Repr
+
+instance : BEq Timestamp where
+  beq x y := x.second == y.second && y.nano == x.nano
+
+instance : Inhabited Timestamp where
+  default := ⟨0, Bounded.LE.mk 0 (by decide), by decide⟩
+
+instance : OfNat Timestamp n where
+  ofNat := by
+    refine ⟨UnitVal.mk n, ⟨0, by decide⟩, ?_⟩
+    simp <;> exact Int.le_total s.val 0 |>.symm
+    exact Int.le_total 0 n
+
+namespace Timestamp
+
+/--
+Fetches the current timestamp from the system.
+-/
+@[extern "lean_get_current_time"]
+opaque now : IO Timestamp
+
+/--
+Converts a `Timestamp` into its equivalent `Second.Offset`.
+-/
+@[inline]
+def toSeconds (t : Timestamp) : Second.Offset :=
+  t.second
+
+/--
+Negates a `Timestamp`, flipping its second and nanosecond values.
+-/
+@[inline]
+def neg (t : Timestamp) : Timestamp := by
+  refine ⟨-t.second, t.nano.neg, ?_⟩
+  cases t.proof with
+  | inl n => exact Or.inr (n.imp Int.neg_le_neg Int.neg_le_neg)
+  | inr n => exact Or.inl (n.imp Int.neg_le_neg Int.neg_le_neg)
+
+/--
+Adds two timestamps together, handling any carry-over in nanoseconds.
+-/
+def add (t₁ t₂ : Timestamp) : Timestamp := by
```

Comment:
I don't think `neg`, `add` and `sub` should be part of the public API for `Timestamp`, they are semantically only meaningful in the context of `Duration`.

Reply:
I changed them to protected so I can use them in `Duration.lean`